src/lib/shim/src/clone.rs

```diff
@@ -0,0 +1,253 @@
+    let blk = unsafe { blk.as_ref().unwrap() };
+    // SAFETY: ensured by caller
+    unsafe { crate::tls_ipc::set(blk) };
+}
+
+/// Execute a native `clone` syscall. The newly created child thread will
+/// resume execution from `ctx`, which should be the point where the managed
+/// code originally made a `clone` syscall (but was intercepted by seccomp).
+///
+/// # Safety
+///
+/// * `ctx` must be dereferenceable, and must be safe for the newly spawned
+/// child thread to restore.
+/// * Other pointers, if non-null, must be safely dereferenceable.
+/// * `child_stack` must be "sufficiently big" for the child thread to run on.
+/// * `tls` if provided must point to correctly initialized thread local storage.
+pub unsafe fn do_clone(ctx: &ucontext, event: &ShimEventAddThreadReq) -> i64 {
+    let flags = event.flags;
+    let ptid: *mut i32 = event.ptid.cast::<i32>().into_raw_mut();
+    let ctid: *mut i32 = event.ctid.cast::<i32>().into_raw_mut();
+    let child_stack: *mut u8 = event.child_stack.cast::<u8>().into_raw_mut();
+    let newtls = event.newtls;
+
+    assert!(
+        !child_stack.is_null(),
+        "clone without a new stack not implemented"
+    );
+
+    // x86-64 calling conventions require a 16-byte aligned stack
+    assert_eq!(
+        child_stack.align_offset(16),
+        0,
+        "clone with unaligned new stack not implemented"
+    );
+
+    // Copy ctx to top of the child stack.
+    // SAFETY: Should still point within stack, assuming it fits.
+    let child_current_rsp = unsafe { child_stack.sub(core::mem::size_of::<sigcontext>()) };
+    let child_current_rsp =
+        unsafe { align_down(child_current_rsp, core::mem::align_of::<sigcontext>()) };
+    let child_sigcontext = child_current_rsp.cast::<sigcontext>();
+    unsafe { core::ptr::write(child_sigcontext, ctx.uc_mcontext) };
+
+    // Update child's copy of context to use the child's stack.
+    let child_sigctx = unsafe { child_sigcontext.as_mut().unwrap() };
+    child_sigctx.rsp = child_stack as u64;
+
+    // Copy child's IPC block to child's stack
+    let child_current_rsp =
+        unsafe { child_current_rsp.sub(core::mem::size_of::<ShMemBlockSerialized>()) };
+    let child_current_rsp = unsafe {
+        align_down(
+            child_current_rsp,
+            core::mem::align_of::<ShMemBlockSerialized>(),
+        )
+    };
+    let child_ipc_blk = child_current_rsp.cast::<ShMemBlockSerialized>();
+    unsafe { core::ptr::write(child_ipc_blk, event.ipc_block) };
+
+    // Ensure stack is 16-aligned so that we can safely make function calls.
+    let child_current_rsp = unsafe { align_down(child_current_rsp, 16) };
+
+    let rv: i64;
+    // SAFETY:
+    // * There's currently no way to tell Rust than an asm block "returns twice",
+    //   so the child does not return from this asm block. It instead *jumps*
+    //   to the point where the clone syscall war originally made.
```

Comment:
since this can create a new thread with the same address space, doesn't this lead to the same issues as [vfork](https://github.com/rust-lang/libc/issues/1596)? does the jump avoid the same memory corruption issue?

Reply:
Yes that's the idea - we're jumping to where the context was created, which should be where the managed code was trying to make a `clone` syscall, which should be a place where it's already safe to return twice. Thanks for the reference  - I was meaning to look that up again and add it here.

Reply:
Expanded this comment and  added the link to the Rust issue.